### PR TITLE
Deduplicate varmap file

### LINKED ISSUE #110: direnv reload takes up to 13 seconds when changing directories
Haven't done much investigation, but it seems that this is not on the daemon side of things, because it happens even when the rebuild occurs in a much shorter amount of time. I suppose that it could also be a direnv issue, but direnv seems to be running relatively quickly for non lorri expressions. Are there any easy ways to debug this?
```
~/.config/taffybar/taffybar master*
❯ time zsh -c 'direnv reload'
zsh -c 'direnv reload'  0.01s user 0.00s system 102% cpu 0.010 total
direnv: loading .envrc
direnv: ([/nix/store/fvvdirp8mf30241alvbz5y5gqcdansx2-direnv-2.20.0-bin/bin/direnv export zsh]) is taking a while to execute. Use CTRL-C to give up.
direnv: export +AR +AR_FOR_TARGET +AS +AS_FOR_TARGET +CC +CC_FOR_TARGET +CONFIG_SHELL +CXX +CXX_FOR_TARGET +GETTEXTDATADIRS +GSETTINGS_SCHEMAS_PATH +HOST_PATH +IN_NIX_SHELL +LD +LD_FOR_TARGET +NIX_BINTOOLS +NIX_BINTOOLS_WRAPPER_x86_64_unknown_linux_gnu_TARGET_HOST +NIX_BINTOOLS_WRAPPER_x86_64_unknown_linux_gnu_TARGET_TARGET +NIX_BUILD_CORES +NIX_BUILD_TOP +NIX_CC +NIX_CC_WRAPPER_x86_64_unknown_linux_gnu_TARGET_HOST +NIX_CC_WRAPPER_x86_64_unknown_linux_gnu_TARGET_TARGET +NIX_CFLAGS_COMPILE +NIX_ENFORCE_NO_NATIVE +NIX_GHC +NIX_GHCPKG +NIX_GHC_DOCDIR +NIX_GHC_LIBDIR +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_LOG_FD +NIX_SSL_CERT_FILE +NIX_STORE +NIX_TARGET_BINTOOLS +NIX_TARGET_CC +NIX_TARGET_CFLAGS_COMPILE +NIX_TARGET_LDFLAGS +NM +NM_FOR_TARGET +OBJCOPY +OBJCOPY_FOR_TARGET +OBJDUMP +OBJDUMP_FOR_TARGET +PKG_CONFIG_PATH +RANLIB +RANLIB_FOR_TARGET +READELF +READELF_FOR_TARGET +SIZE +SIZE_FOR_TARGET +SOURCE_DATE_EPOCH +SSL_CERT_FILE +STRINGS +STRINGS_FOR_TARGET +STRIP +STRIP_FOR_TARGET +TARGET_AR +TARGET_AS +TARGET_CC +TARGET_CXX +TARGET_LD +TARGET_NM +TARGET_OBJCOPY +TARGET_OBJDUMP +TARGET_RANLIB +TARGET_READELF +TARGET_SIZE +TARGET_STRINGS +TARGET_STRIP +buildInputs +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +extraClosure +installPhase +name +nativeBuildInputs +origArgs +origBuilder +origExtraClosure +origOutputs +origPATH +origSystem +out +outputs +patches +phases +preHook +propagatedBuildInputs +propagatedNativeBuildInputs +shell +stdenv +strictDeps +system ~GDK_PIXBUF_MODULE_FILE ~LANG ~LOCALE_ARCHIVE ~PATH

~/.config/taffybar/taffybar master* 13s
❯                        ~/.config/taffybar/taffybar
direnv: loading .envrc
```

### PR DESCRIPTION
**Pull request by IvanMalison**
Fixes #110

### CODE REVIEW

**Comment by IvanMalison** [comment]
Hmmm, Looks like this didn't quite do what I intended. Thankfully there are tests that are now failing that caught that ther is an issue.

**Comment by IvanMalison** [comment]
Okay, so the issue is that bash strings can't ever contain null bytes, so they get dropped when they are assigned to a variable. The easy thing to do here is to just redo the printf when actually writing stuff out.

The check still actually works because the null bytes get dropped from both append_string and current_contents, but it feels a little hacky that things are done this way. I dont really have a better idea though so I'm just going to leave as is.

Maybe someone with a bit more bash fu than me has a better idea.

**Comment by IvanMalison** [comment]
I suppose something like:

```
```

**Comment by IvanMalison** [comment]
Okay, this nasty while loop thing is the most correct thing I could come up with. It's not pretty, but it gets the job done. @grahamc What do you think?

File: src/logged-evaluation.nix [hunk 1]
```diff
@@ -88,7 +88,31 @@ let
...
+            elif [ $modded -eq 2 ]; then
+              match_delimeter="$value"
+              if [[ $match_varname == $varname &&
+                    $match_delimeter == $delimiter ]]; then
+                already_present=1
+                break
+              fi
+            fi
+            index=$(($index + 1))
+          done < $out/varmap-v1
```
[the remark this answers is not in the document]
Comment (IvanMalison) [reply src/logged-evaluation.nix]: We touch the file here:
https://github.com/target/lorri/blob/938085e2b6bae501e5b45f5ff6979a0312ca74b1/src/logged-evaluation.nix#L126

but if you want to be extra paranoid about it we could do it again in the function, I suppose.

File: src/logged-evaluation.nix [hunk 2]
```diff
@@ -88,7 +88,31 @@ let
...
           local varname=$2
-          printf 'append\0%s\0%s\0' "$varname" "$delimiter" >> "$out/varmap-v1"
+
+          local already_present=0
+          local index=0
+          local match_varname
+          local match_delimiter
+
+          while read -d $'\0' value ; do
+            local modded=$(( $index % 3 ))
```
[the remark this answers is not in the document]
Comment (IvanMalison) [reply src/logged-evaluation.nix]: ah much nicer. Guess it didn't occur to me to do that sort of trick.

**Comment by IvanMalison** [comment]
> Looks great! I've just push a commit to your branch which tests the speed of the hook, to prevent this regression in the future.

Nice. Good call on the test!

**Comment by grahamc** [comment]
Yeah! I kind of wished there was a way to assert the contents of the varmap file, but it looks like the current option is to assert the timing of the result. Oh well. The timing is the real bug report anyway :)

**Comment by grahamc** [comment]
Thank you!

**Pull request merged**